### PR TITLE
Cleanup and throw exception if bind failed

### PR DESCRIPTION
**Pull request by AlbumenJ**
## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/server/Server.java [hunk 1]
```diff
@@ -115,17 +115,33 @@ protected void initChannel(Channel ch) throws Exception {
                 ));
             }
         });
-        try {
-            if (StringUtils.isBlank(host)) {
-                serverBootstrap.bind(port).sync();
+        Throwable lastError = null;
+        for (int i = 0; i < 10; i++) {
```
Comment (CrazyHZM) [inline dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/server/Server.java:119]: I recommend adding a server-failed-retry-count  and  server-failed-retry-interval configuration

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/server/Server.java (hunk 1 above)
Comment (AlbumenJ) [reply dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/server/Server.java:119]: ok

Comment (AlbumenJ) [reply dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/server/Server.java]: done

File: dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyPortUnificationServer.java [hunk 2]
```diff
@@ -138,9 +139,31 @@ protected void initChannel(SocketChannel ch) throws Exception {
             bindIp = ANYHOST_VALUE;
         }
         InetSocketAddress bindAddress = new InetSocketAddress(bindIp, bindPort);
-        ChannelFuture channelFuture = bootstrap.bind(bindAddress);
-        channelFuture.syncUninterruptibly();
-        channel = channelFuture.channel();
+        Throwable lastError = null;
+        int retryTimes = getUrl().getParameter(Constants.BIND_RETRY_TIMES, 10);
+        int retryInterval = getUrl().getParameter(Constants.BIND_RETRY_INTERVAL, 3000);
+        for (int i = 0; i < retryTimes; i++) {
```
Comment (chickenlj) [inline dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyPortUnificationServer.java:145]: Why would we need binding retry, what's the problem with throwing exception directly on binding failure.

Comment (AlbumenJ) [reply dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyPortUnificationServer.java:145]: Such 20880, 22222... are in `ip_local_port_range` and can be used as a tcp client port. If there is someone requesting remoting using 20880 as client port(allocated by kernel), Dubbo will bind failed.

**Review by CrazyHZM — APPROVED** [review]

File: dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/protocol/QosProtocolWrapper.java [hunk 3]
```diff
@@ -128,13 +130,12 @@ private void startQosServer(URL url) {
...
             server.setAnonymousAllowCommands(anonymousAllowCommands);
-            server.start();
+            server.start(qosCheck);
 
         } catch (Throwable throwable) {
             logger.warn(QOS_FAILED_START_SERVER, "", "", "Fail to start qos server: ", throwable);
-            boolean qosCheck = url.getParameter(QOS_CHECK, false);
             if (qosCheck) {
-                throw new IllegalStateException("Fail to start qos server: " + throwable.getMessage(), throwable);
+                throw new RpcException(throwable);
```
Comment (chickenlj) [inline dubbo-plugin/dubbo-qos/src/main/java/org/apache/dubbo/qos/protocol/QosProtocolWrapper.java:138]: Unnecessary wrap, how about throw it directly.

**Review by chickenlj — APPROVED** [review]
LGTM.

**Pull request merged**